### PR TITLE
feat: Add experimental async transport (port of PR #4572)

### PR DESCRIPTION
**Pull request by BYK**
Add an experimental async transport using httpcore's async backend,
enabled via `_experiments={"transport_async": True}`.

This is a manual port of PR #4572 (originally merged into `potel-base`)
onto the current `master` branch.

## Key changes

- **transport.py**: Refactor `BaseHttpTransport` into `HttpTransportCore`
  (shared base) + `BaseHttpTransport` (sync) + `AsyncHttpTransport`
  (async, conditional on `httpcore[asyncio]`). Extract shared helpers:
  `_handle_request_error`, `_handle_response`, `_update_headers`,
  `_prepare_envelope`. Update `make_transport()` to detect the
  `transport_async` experiment.

- **worker.py**: Add `Worker` ABC base class and `AsyncWorker`
  implementation using `asyncio.Queue` / `asyncio.Task`.

- **client.py**: Add `close_async()` / `flush_async()` with async-vs-sync
  transport detection. Extract `_close_components()` / `_flush_components()`.

- **api.py**: Expose `flush_async()` as a public API.

- **integrations/asyncio.py**: Patch `loop.close` to flush pending events
  before shutdown. Skip span wrapping for internal Sentry tasks.

- **utils.py**: Add `is_internal_task()` / `mark_sentry_task_internal()`
  via ContextVar for async task filtering.

- **setup.py**: Add `"asyncio"` extras_require (`httpcore[asyncio]==1.*`).

- **config.py / tox.ini**: Widen anyio to `>=3,<5` for httpx and FastAPI.

## Notes

- `tox.ini` was manually edited (the generation script requires a
  free-threaded Python interpreter). A full regeneration should be done
  before merge.

Refs: GH-4568

### CODE REVIEW

File: sentry_sdk/integrations/asyncio.py [hunk 1]
```diff
@@ -156,6 +220,7 @@ def __init__(self, task_spans: bool = True) -> None:
     @staticmethod
     def setup_once() -> None:
         patch_asyncio()
+        patch_loop_close()
```
[the remark this answers is not in the document]
Comment (sentrivana) [reply sentry_sdk/integrations/asyncio.py:223]: `enable_asyncio_integration` literally calls `setup_once`, so it'll get any changes we make to `setup_once` automatically

File: sentry_sdk/client.py [hunk 2]
```diff
@@ -1009,19 +1037,46 @@ def close(
...
+                and hasattr(self.transport, "loop")
+            ):
+                logger.debug(
+                    "close_async() used with non-async transport, aborting. Please use close() instead."
+                )
+                return
+            await self.flush_async(timeout=timeout, callback=callback)
+            self._close_components()
+            kill_task = self.transport.kill()  # type: ignore
+            if kill_task is not None:
```
Comment (sentrivana) [inline sentry_sdk/client.py:1078]: When can `kill_task` be `None`?

File: sentry_sdk/client.py [hunk 3]
```diff
@@ -1035,23 +1090,59 @@ def flush(
         :param callback: Is invoked with the number of pending events and the configured timeout.
         """
         if self.transport is not None:
+            if isinstance(self.transport, AsyncHttpTransport) and hasattr(
+                self.transport, "loop"
+            ):
```
Comment (sentrivana) [inline sentry_sdk/client.py]: Why is the `hasattr(self.transport, "loop")` part important here?

File: sentry_sdk/client.py [hunk 4]
```diff
@@ -1035,23 +1090,59 @@ def flush(
...
+
+        :param timeout: Wait for at most `timeout` seconds. If no `timeout` is provided, the `shutdown_timeout` option value is used.
+
+        :param callback: Is invoked with the number of pending events and the configured timeout.
+        """
+        if self.transport is not None:
+            if not (
+                isinstance(self.transport, AsyncHttpTransport)
+                and hasattr(self.transport, "loop")
+            ):
```
Comment (sentrivana) [inline sentry_sdk/client.py]: https://github.com/getsentry/sentry-python/pull/5646/changes#r2986964659

File: AGENTS.md [hunk 5]
```diff
@@ -67,3 +67,18 @@ Do NOT edit these directly — modify source scripts instead:
...
+* **Consola prompt cancel returns truthy Symbol, not false**: When a user cancels a \`consola\` / \`@clack/prompts\` confirmation prompt (Ctrl+C), the return value is \`Symbol(clack:cancel)\`, not \`false\`. Since Symbols are truthy in JavaScript, checking \`!confirmed\` will be \`false\` and the code falls through as if the user confirmed. Fix: use \`confirmed !== true\` (strict equality) instead of \`!confirmed\` to correctly handle cancel, false, and any other non-true values.
+
+<!-- lore:019cc303-e397-75b9-9762-6f6ad108f50a -->
+* **Zod z.coerce.number() converts null to 0 silently**: Zod gotchas in this codebase: (1) \`z.coerce.number()\` passes input through \`Number()\`, so \`null\` silently becomes \`0\`. Be aware if \`null\` vs \`0\` distinction matters. (2) Zod v4 \`.default({})\` short-circuits — it returns the default value without parsing through inner schema defaults. So \`.object({ enabled: z.boolean().default(true) }).default({})\` returns \`{}\`, not \`{ enabled: true }\`. Fix: provide fully-populated default objects. This affected nested config sections in src/config.ts during the v3→v4 upgrade.
+<!-- End lore-managed section -->
```
Comment (sentrivana) [inline AGENTS.md]: Pls remove

File: tests/integrations/asyncio/test_asyncio.py [hunk 6]
```diff
@@ -564,3 +573,428 @@ async def test_delayed_enable_integration_after_disabling(sentry_init, capture_e
...
+    finally:
+        if not loop.is_closed():
+            loop.close()
+        if original_loop:
+            asyncio.set_event_loop(original_loop)
+
+
+# ============================================================================
+# patch_loop_close edge case tests
+# ============================================================================
```
Comment (sentrivana) [inline tests/integrations/asyncio/test_asyncio.py]: Can we remove this

File: tests/integrations/asyncio/test_asyncio.py [hunk 7]
```diff
@@ -564,3 +573,428 @@ async def test_delayed_enable_integration_after_disabling(sentry_init, capture_e
...
+    with patch("asyncio.get_running_loop", return_value=mock_loop):
+        patch_loop_close()
+
+    # close should have been replaced
+    assert mock_loop._sentry_flush_patched is True
+
+
+# ============================================================================
+# _create_task_with_factory tests
+# ============================================================================
```
Comment (sentrivana) [inline tests/integrations/asyncio/test_asyncio.py]: Please remove

File: tests/integrations/asyncio/test_asyncio.py [hunk 8]
```diff
@@ -564,3 +573,428 @@ async def test_delayed_enable_integration_after_disabling(sentry_init, capture_e
...
+    mock_loop = MagicMock()
+    mock_loop._sentry_flush_patched = False
+    # Make getattr return False for _sentry_flush_patched
+    type(mock_loop)._sentry_flush_patched = False
+
+    with patch("asyncio.get_running_loop", return_value=mock_loop):
+        patch_loop_close()
+
+    # close should have been replaced
+    assert mock_loop._sentry_flush_patched is True
```
Comment (sentrivana) [inline tests/integrations/asyncio/test_asyncio.py]: I don't see the benefit of this test case when we already have `test_loop_close_patching`

File: tests/integrations/asyncio/test_asyncio.py [hunk 9]
```diff
@@ -564,3 +573,428 @@ async def test_delayed_enable_integration_after_disabling(sentry_init, capture_e
...
+    result = _create_task_with_factory(orig_factory, mock_loop, mock_coro)
+
+    orig_factory.assert_called_once()
+    MockTask.assert_called_once_with(mock_coro, loop=mock_loop)
+    assert result == MockTask.return_value
+
+
+# ============================================================================
+# _sentry_task_factory internal task detection tests
+# ============================================================================
```
Comment (sentrivana) [inline tests/integrations/asyncio/test_asyncio.py]: Please remove (I'll stop commenting on these now, but same applies to other top-level comments like this)

File: tests/integrations/asyncio/test_asyncio.py [hunk 10]
```diff
@@ -564,3 +573,428 @@ async def test_delayed_enable_integration_after_disabling(sentry_init, capture_e
...
+
+
+@minimum_python_39
+@pytest.mark.asyncio(loop_scope="module")
+async def test_sentry_task_factory_skips_internal_tasks(sentry_init):
```
Comment (sentrivana) [inline tests/integrations/asyncio/test_asyncio.py]: What's the difference to `test_internal_tasks_not_wrapped`?

File: tests/integrations/asyncio/test_asyncio.py [hunk 11]
```diff
@@ -564,3 +573,428 @@ async def test_delayed_enable_integration_after_disabling(sentry_init, capture_e
...
+# is_internal_task / mark_sentry_task_internal utility tests
+# ============================================================================
+
+
+@minimum_python_38
+def test_is_internal_task_default():
+    """Test is_internal_task returns False by default."""
+    from sentry_sdk.utils import is_internal_task
+
+    assert is_internal_task() is False
```
Comment (sentrivana) [inline tests/integrations/asyncio/test_asyncio.py]: This test is unnecessary. We should test the behavior, ideally end-to-end, not the value of a single variable

File: tests/integrations/asyncio/test_asyncio.py [hunk 12]
```diff
@@ -564,3 +573,428 @@ async def test_delayed_enable_integration_after_disabling(sentry_init, capture_e
...
+    """Test nested mark_sentry_task_internal restores correctly."""
+    from sentry_sdk.utils import is_internal_task, mark_sentry_task_internal
+
+    assert is_internal_task() is False
+    with mark_sentry_task_internal():
+        assert is_internal_task() is True
+        with mark_sentry_task_internal():
+            assert is_internal_task() is True
+        assert is_internal_task() is True
+    assert is_internal_task() is False
```
Comment (sentrivana) [inline tests/integrations/asyncio/test_asyncio.py]: Same comment as above: We should test the complete behavior rather than what the specific implementation is doing. Let's either get rid of these or make them create actual tasks and check the resulting spans

File: tests/integrations/asyncio/test_asyncio.py [hunk 13]
```diff
@@ -564,3 +573,428 @@ async def test_delayed_enable_integration_after_disabling(sentry_init, capture_e
...
+    from sentry_sdk.utils import is_internal_task, mark_sentry_task_internal
+
+    assert is_internal_task() is False
+    try:
+        with mark_sentry_task_internal():
+            assert is_internal_task() is True
+            raise ValueError("test exception")
+    except ValueError:
+        pass
+    assert is_internal_task() is False
```
Comment (sentrivana) [inline tests/integrations/asyncio/test_asyncio.py]: This is a good test, but again it tests implementation details. Can we change it to actually create some tasks and check that they are captured correctly?

File: tests/integrations/asyncio/test_asyncio.py [hunk 14]
```diff
@@ -564,3 +573,428 @@ async def test_delayed_enable_integration_after_disabling(sentry_init, capture_e
...
+    from sentry_sdk.integrations.asyncio import patch_loop_close
+
+    async def _inner():
+        loop = asyncio.get_running_loop()
+        if hasattr(loop, "_sentry_flush_patched"):
+            delattr(loop, "_sentry_flush_patched")
+        patch_loop_close()
+        assert getattr(loop, "_sentry_flush_patched", False) is True
+
+    asyncio.run(_inner())
```
Comment (sentrivana) [inline tests/integrations/asyncio/test_asyncio.py]: This doesn't appear to be testing anything that makes sense to test, please no tests just for the sake of increasing coverage

File: tests/integrations/asyncio/test_asyncio.py [hunk 15]
```diff
@@ -564,3 +573,428 @@ async def test_delayed_enable_integration_after_disabling(sentry_init, capture_e
...
+        assert await task2 == "hello"
+
+        # Factory returns None — should fall back to Task()
+        def none_factory(lp, coro, **kw):
+            return None
+
+        task3 = _create_task_with_factory(none_factory, loop, dummy())
+        assert await task3 == "hello"
+
+    asyncio.run(_inner())
```
Comment (sentrivana) [inline tests/integrations/asyncio/test_asyncio.py]: https://github.com/getsentry/sentry-python/pull/5646/changes#r2987175096

File: tests/integrations/asyncio/test_asyncio.py [hunk 16]
```diff
@@ -564,3 +573,428 @@ async def test_delayed_enable_integration_after_disabling(sentry_init, capture_e
...
+        assert is_internal_task() is False
+        with mark_sentry_task_internal():
+            assert is_internal_task() is True
+            # Nested
+            with mark_sentry_task_internal():
+                assert is_internal_task() is True
+            assert is_internal_task() is True
+        assert is_internal_task() is False
+
+    asyncio.run(_inner())
```
Comment (sentrivana) [inline tests/integrations/asyncio/test_asyncio.py]: https://github.com/getsentry/sentry-python/pull/5646/changes#r2987158984

File: tests/integrations/asyncio/test_asyncio.py [hunk 17]
```diff
@@ -564,3 +573,428 @@ async def test_delayed_enable_integration_after_disabling(sentry_init, capture_e
...
+
+
+@minimum_python_38
+def test_sentry_task_factory_integration_sync():
```
Comment (sentrivana) [inline tests/integrations/asyncio/test_asyncio.py]: This looks like one of the basic tests above, I don't see the added benefit

File: tests/test_client.py [hunk 18]
```diff
@@ -1585,3 +1598,640 @@ def test_keep_alive(env_value, arg_value, expected_value):
...
+        await client.flush_async(timeout=1.0)
+
+    assert flush_awaited == [True]
+    await client.close_async()
+
+
+@skip_under_gevent
+@pytest.mark.asyncio
+@pytest.mark.skipif(not PY38, reason="Async client methods require Python 3.8+")
+async def test_close_async_awaits_kill_task():
```
Comment (sentrivana) [inline tests/test_client.py]: Isn't this the same as `test_close_async_with_async_transport`

File: tests/test_client.py [hunk 19]
```diff
@@ -1585,3 +1598,640 @@ def test_keep_alive(env_value, arg_value, expected_value):
...
+
+    client.session_flusher = mock.MagicMock()
+    client.log_batcher = None
+    client.metrics_batcher = None
+    client.span_batcher = None
+
+    # Should not raise
+    client._flush_components()
+    client.session_flusher.flush.assert_called_once()
+    client.close()
```
Comment (sentrivana) [inline tests/test_client.py]: These test implementation details, please remove

File: tests/test_transport.py [hunk 20]
```diff
@@ -841,3 +856,3200 @@ def test_record_lost_event_transaction_item(capturing_server, make_client, span_
         "reason": "test",
         "quantity": span_count + 1,
     } in discarded_events
+
```
Comment (sentrivana) [inline tests/test_transport.py:879]: The tests in this file are currently not reviewable for me. It's too many.

I have some general points though:
- Can we get rid of the implementation details tests, and instead aim for behavior tests?
- Since there is precedent with the sync transport tests, I think it'd be enough to copy those and adapt them for the async transport.

File: tox.ini [hunk 21]
```diff
@@ -385,14 +384,17 @@ deps =
...
 
     mypy: -r requirements-linting.txt
     mypy: werkzeug<2.3.0
+    mypy: httpcore[asyncio]
     ruff: -r requirements-linting.txt
 
     # === Common ===
     py3.8-common: hypothesis
     common: pytest-asyncio
+    common: httpcore[asyncio]
```
Comment (sentrivana) [inline tox.ini:397]: `tox.ini` is automatically generated so these changes won't persist:
- The `common` test suite and `linters` and `mypy` related changes should be made in the template in `scripts/populate_tox/tox.jinja`.
- The integration test suite related changes (`httpx`, `fastapi`) should be made in `scripts/populate_tox/config.py`.
- You can then regenerate `tox.ini` via `scripts/generate-test-files.sh` -- it needs a free-threading build of Python though, so if there's any trouble, happy to take care of it.

File: sentry_sdk/transport.py [hunk 22]
```diff
@@ -505,6 +536,74 @@ def _serialize_envelope(
...
+        }
+
+        socket_options = (
+            self.options["socket_options"]
+            if self.options["socket_options"] is not None
+            else []
+        )
+
+        used_options = {(o[0], o[1]) for o in socket_options}
+        for default_option in KEEP_ALIVE_SOCKET_OPTIONS:
```
Comment (sentrivana) [inline sentry_sdk/transport.py:557]: Why are we always adding these, even when `keep_alive=False`?

File: sentry_sdk/client.py [hunk 23]
```diff
@@ -999,6 +1009,32 @@ def get_integration(
 
         return self.integrations.get(integration_name)
 
+    def _is_async_transport(self) -> bool:
```
Comment (sentrivana) [inline sentry_sdk/client.py]:
```suggestion
    def _has_async_transport(self) -> bool:
```

Since this is a method on the client, `has` fits better

File: sentry_sdk/client.py [hunk 24]
```diff
@@ -1009,19 +1045,41 @@ def close(
...
+        timeout: "Optional[float]" = None,
+        callback: "Optional[Callable[[int, float], None]]" = None,
+    ) -> None:
+        """
+        Asynchronously close the client and shut down the transport. Arguments have the same
+        semantics as :py:meth:`Client.flush_async`.
+        """
+        if self.transport is not None:
+            if not self._is_async_transport():
+                logger.debug(
```
Comment (sentrivana) [inline sentry_sdk/client.py:1071]: Can we make this a `warnings.warn`, so that it's always logged once?

File: sentry_sdk/client.py [hunk 25]
```diff
@@ -1035,23 +1093,54 @@ def flush(
         :param callback: Is invoked with the number of pending events and the configured timeout.
         """
         if self.transport is not None:
+            if self._is_async_transport():
+                logger.warning(
+                    "flush() used with AsyncHttpTransport. Please use flush_async() instead."
+                )
```
Comment (sentrivana) [inline sentry_sdk/client.py]: Can we do `warnings.warn` here too?

File: sentry_sdk/transport.py [hunk 26]
```diff
@@ -15,13 +16,37 @@
...
+    import h2  # noqa: F401
+
+    HTTP2_ENABLED = httpcore is not None
+except ImportError:
+    HTTP2_ENABLED = False
+
+try:
+    import anyio  # noqa: F401
+
+    ASYNC_TRANSPORT_ENABLED = httpcore is not None
```
Comment (sentrivana) [inline sentry_sdk/transport.py]:
```suggestion
    ASYNC_TRANSPORT_AVAILABLE = httpcore is not None
```

`ENABLED` evokes that it's actually active, which it might not be since the `init` option might be `False`

File: tests/test_client.py [hunk 27]
```diff
@@ -1585,3 +1596,315 @@ def test_keep_alive(env_value, arg_value, expected_value):
...
+    client = Client(
+        "https://foo@sentry.io/123",
+        _experiments={"transport_async": True},
+        integrations=[AsyncioIntegration()],
+    )
+    assert isinstance(client.transport, AsyncHttpTransport)
+
+    # close_async should call kill() and await the returned task
+    await client.close_async(timeout=1.0)
+    assert client.transport is None
```
Comment (sentrivana) [inline tests/test_client.py]:
```suggestion
```
As far as I can tell this is identical to `test_close_async_with_async_transport`. Please remove.

File: tests/test_transport.py [hunk 28]
```diff
@@ -841,3 +877,1278 @@ def test_record_lost_event_transaction_item(capturing_server, make_client, span_
...
+
+def test_handle_unexpected_status_invokes_handle_request_error(
```
Comment (sentrivana) [inline tests/test_transport.py]: Please remove or change this one, it's mocking everything, making it too dependent on implementation details. If we want to test this, it should use the capturing server fixture, which supports custom error codes with `capturing_server.respond_with(code=500)`, and we shouldn't check `_handle_request_error`, but rather something like `record_lost_event`. There's prior art in `test_data_category_limits` or `test_envelope_too_large_response`.

File: tests/test_transport.py [hunk 29]
```diff
@@ -841,3 +877,1278 @@ def test_record_lost_event_transaction_item(capturing_server, make_client, span_
...
+    envelope = Envelope()
+    envelope.add_item(mock.MagicMock())  # Simple mock item
+    envelope.add_item(mock.MagicMock())  # Another mock item
+
+    transport._handle_request_error(envelope=envelope, loss_reason="connection_error")
+
+    assert len(calls) == 3
+    assert calls[0] == ("on_dropped_event", "connection_error")
+    assert calls[1][0:2] == ("record_lost_event", "network_error")
+    assert calls[2][0:2] == ("record_lost_event", "network_error")
```
Comment (sentrivana) [inline tests/test_transport.py]:
```suggestion
```

I don't see the added benefit of this test, please remove

File: tests/test_transport.py [hunk 30]
```diff
@@ -841,3 +877,1278 @@ def test_record_lost_event_transaction_item(capturing_server, make_client, span_
...
+@skip_under_gevent
+@pytest.mark.asyncio
+@pytest.mark.skipif(not PY38, reason="Async transport requires Python 3.8+")
+async def test_async_transport_background_thread_capture(
+    capturing_server, make_client, caplog
+):
+    """Test capture_envelope from background threads uses run_coroutine_threadsafe"""
+    caplog.set_level(logging.DEBUG)
+    experiments = {"transport_async": True}
+    client = make_client(_experiments=experiments, integrations=[AsyncioIntegration()])
```
Comment (sentrivana) [inline tests/test_transport.py]:
```suggestion
    client = make_client(_experiments={"transport_async": True}, integrations=[AsyncioIntegration()])
```

File: tests/test_transport.py [hunk 31]
```diff
@@ -841,3 +877,1278 @@ def test_record_lost_event_transaction_item(capturing_server, make_client, span_
...
+    # Therefore, we need to explicitly close the client to clean up the worker task
+    assert any("Sending envelope" in record.msg for record in caplog.records) == debug
+    if client_flush_method == "flush":
+        await client.close_async(timeout=2.0)
+
+
+@skip_under_gevent
+@pytest.mark.asyncio
+@pytest.mark.skipif(not PY38, reason="Async transport requires Python 3.8+")
+async def test_async_transport_background_thread_capture(
```
Comment (sentrivana) [inline tests/test_transport.py]: Unsure about this test, in which scenarios could there be issues when capturing from a thread with the async transport?

File: tests/test_transport.py [hunk 32]
```diff
@@ -841,3 +877,1278 @@ def test_record_lost_event_transaction_item(capturing_server, make_client, span_
...
+@skip_under_gevent
+@pytest.mark.asyncio
+@pytest.mark.skipif(not PY38, reason="Async transport requires Python 3.8+")
+async def test_async_transport_event_loop_closed_scenario(
+    capturing_server, make_client, caplog
+):
+    """Test behavior when trying to capture after event loop context ends"""
+    caplog.set_level(logging.DEBUG)
+    experiments = {"transport_async": True}
+    client = make_client(_experiments=experiments, integrations=[AsyncioIntegration()])
```
Comment (sentrivana) [inline tests/test_transport.py]:
```suggestion
    client = make_client(_experiments={"transport_async": True}, integrations=[AsyncioIntegration()])
```

File: tests/test_transport.py [hunk 33]
```diff
@@ -841,3 +877,1278 @@ def test_record_lost_event_transaction_item(capturing_server, make_client, span_
...
+    sentry_sdk.get_global_scope().set_client(client)
+    original_loop = client.transport.loop
+
+    with mock.patch("asyncio.get_running_loop", side_effect=RuntimeError("no loop")):
+        with mock.patch.object(client.transport.loop, "is_running", return_value=False):
```
Comment (sentrivana) [inline tests/test_transport.py]: Can we, instead of mocking, actually close the event loop and verify that the transport handles this gracefully?

File: tests/test_transport.py [hunk 34]
```diff
@@ -841,3 +877,1278 @@ def test_record_lost_event_transaction_item(capturing_server, make_client, span_
...
+@skip_under_gevent
+@pytest.mark.asyncio
+@pytest.mark.skipif(not PY38, reason="Async transport requires Python 3.8+")
+async def test_async_transport_concurrent_requests(
+    capturing_server, make_client, caplog
+):
+    """Test multiple simultaneous envelope submissions"""
+    caplog.set_level(logging.DEBUG)
+    experiments = {"transport_async": True}
+    client = make_client(_experiments=experiments, integrations=[AsyncioIntegration()])
```
Comment (sentrivana) [inline tests/test_transport.py]:
```suggestion
    client = make_client(_experiments={"transport_async": True}, integrations=[AsyncioIntegration()])
```

File: tests/test_transport.py [hunk 35]
```diff
@@ -841,3 +877,1278 @@ def test_record_lost_event_transaction_item(capturing_server, make_client, span_
...
+@pytest.mark.filterwarnings("ignore::pytest.PytestUnraisableExceptionWarning")
+async def test_async_worker_is_alive_after_start():
+    """Test is_alive returns True after start() in a running loop."""
+    from sentry_sdk.worker import AsyncWorker
+
+    worker = AsyncWorker()
+    worker.start()
+    assert worker.is_alive is True
+    worker.kill()
+    await asyncio.sleep(0)  # Allow cancelled tasks to be cleaned up
```
Comment (sentrivana) [inline tests/test_transport.py]:
```suggestion
```

These are testing internal implementation details of the worker

File: tests/test_transport.py [hunk 36]
```diff
@@ -841,3 +877,1278 @@ def test_record_lost_event_transaction_item(capturing_server, make_client, span_
...
+    assert mock_task not in worker._active_tasks
+    worker.kill()
+    await asyncio.sleep(0)  # Allow cancelled tasks to be cleaned up
+
+
+@skip_under_gevent
+@pytest.mark.asyncio
+@pytest.mark.skipif(not PY38, reason="AsyncWorker requires Python 3.8+")
+@pytest.mark.filterwarnings("ignore::pytest.PytestUnraisableExceptionWarning")
+async def test_async_worker_ensure_task_calls_start():
```
Comment (sentrivana) [inline tests/test_transport.py]: Let's remove all the `test_async_worker_*` tests please. They all test implementation details.

File: tests/test_transport.py [hunk 37]
```diff
@@ -841,3 +877,1278 @@ def test_record_lost_event_transaction_item(capturing_server, make_client, span_
...
+    mock_logger.warning.assert_any_call(
+        "HTTP/2 transport is not supported with async transport. "
+        "Ignoring transport_http2 experiment."
+    )
+    await transport._pool.aclose()
+
+
+@skip_under_gevent
+@pytest.mark.skipif(not PY38, reason="Async transport requires Python 3.8+")
+def test_make_transport_async_not_enabled():
```
Comment (sentrivana) [inline tests/test_transport.py]: Let's remove all `test_make_transport_*` tests, they're testing implementation details.

File: tests/test_transport.py [hunk 38]
```diff
@@ -841,3 +877,1278 @@ def test_record_lost_event_transaction_item(capturing_server, make_client, span_
...
+        ),
+    )
+
+    class MockResponse:
+        status = 413
+        headers = {}
+        data = b"entity too large"
+
+    transport._handle_response(MockResponse(), envelope=None)
+    assert error_calls == [("status_413", "send_error")]
```
Comment (sentrivana) [inline tests/test_transport.py]:
```suggestion
```

There's `test_envelope_too_large_response` already which tests this case.

File: tests/test_transport.py [hunk 39]
```diff
@@ -841,3 +877,1278 @@ def test_record_lost_event_transaction_item(capturing_server, make_client, span_
...
+        "on_dropped_event",
+        lambda reason: dropped_reasons.append(reason),
+    )
+
+    class MockResponse:
+        status = 429
+        headers = {"Retry-After": "60"}
+
+    transport._handle_response(MockResponse(), envelope=None)
+    assert dropped_reasons == ["status_429"]
```
Comment (sentrivana) [inline tests/test_transport.py]:
```suggestion
```

There are other test cases testing for 429s

File: tests/test_transport.py [hunk 40]
```diff
@@ -841,3 +877,1278 @@ def test_record_lost_event_transaction_item(capturing_server, make_client, span_
...
+        ),
+    )
+
+    class MockResponse:
+        status = 500
+        headers = {}
+        data = b"server error"
+
+    transport._handle_response(MockResponse(), envelope=None)
+    assert error_calls == ["status_500"]
```
Comment (sentrivana) [inline tests/test_transport.py]:
```suggestion
```

This is identical to `test_handle_unexpected_status_invokes_handle_request_error`

File: tests/test_transport.py [hunk 41]
```diff
@@ -841,3 +877,1278 @@ def test_record_lost_event_transaction_item(capturing_server, make_client, span_
...
+
+def test_update_headers(make_client):
+    """Test _update_headers adds auth and user-agent headers."""
+    client = make_client()
+    transport = client.transport
+    headers = {}
+    transport._update_headers(headers)
+    assert "User-Agent" in headers
+    assert "X-Sentry-Auth" in headers
+    assert "sentry.python" in headers["User-Agent"]
```
Comment (sentrivana) [inline tests/test_transport.py]:
```suggestion
```
Implementation details or superfluous, please remove

File: tests/test_transport.py [hunk 42]
```diff
@@ -841,3 +877,1278 @@ def test_record_lost_event_transaction_item(capturing_server, make_client, span_
...
+
+    envelope = Envelope()
+    envelope.add_event({"type": "error", "message": "test"})
+
+    result = transport._prepare_envelope(envelope)
+    assert result is not None
+    env, body, headers = result
+    # Should have the original item + client report
+    types = [item.type for item in env.items]
+    assert "client_report" in types
```
Comment (sentrivana) [inline tests/test_transport.py]:
```suggestion
```

Implementation details, please remove

File: tests/test_transport.py [hunk 43]
```diff
@@ -841,3 +877,1278 @@ def test_record_lost_event_transaction_item(capturing_server, make_client, span_
...
+            (b"x-sentry-rate-limits", b"60:error:organization"),
+        ]
+
+    val = client.transport._get_header_value(MockResponse(), "X-Sentry-Rate-Limits")
+    assert val == "60:error:organization"
+
+    val = client.transport._get_header_value(MockResponse(), "Nonexistent-Header")
+    assert val is None
+
+    await client.close_async()
```
Comment (sentrivana) [inline tests/test_transport.py]:
```suggestion
```

Implementation details, please remove

File: tests/test_transport.py [hunk 44]
```diff
@@ -841,3 +877,1278 @@ def test_record_lost_event_transaction_item(capturing_server, make_client, span_
...
+
+    with mock.patch.object(client.transport, "loop") as mock_loop:
+        mock_loop.is_running.return_value = False
+        with mock.patch("sentry_sdk.transport.logger") as mock_logger:
+            client.transport.capture_envelope(envelope)
+            mock_logger.warning.assert_called_with(
+                "Async Transport is not running in an event loop."
+            )
+
+    await client.close_async()
```
Comment (sentrivana) [inline tests/test_transport.py]:
```suggestion
```

This is almost the same as `test_async_transport_event_loop_closed_scenario`. Please remove.

File: tests/test_transport.py [hunk 45]
```diff
@@ -841,3 +877,1278 @@ def test_record_lost_event_transaction_item(capturing_server, make_client, span_
...
+        "https://foo@sentry.io/123",
+        _experiments={"transport_async": True},
+        integrations=[AsyncioIntegration()],
+    )
+    assert isinstance(client.transport, AsyncHttpTransport)
+
+    kill_task = client.transport.kill()
+    assert kill_task is not None
+    assert isinstance(kill_task, asyncio.Task)
+    await kill_task
```
Comment (sentrivana) [inline tests/test_transport.py]:
```suggestion
```

Implementation details, please remove

File: tests/test_transport.py [hunk 46]
```diff
@@ -841,3 +877,1278 @@ def test_record_lost_event_transaction_item(capturing_server, make_client, span_
...
+    client = Client(
+        "https://foo@sentry.io/123",
+        _experiments={"transport_async": True},
+        integrations=[AsyncioIntegration()],
+    )
+    assert isinstance(client.transport, AsyncHttpTransport)
+
+    result = client.transport.flush(timeout=0)
+    assert result is None
+    await client.close_async()
```
Comment (sentrivana) [inline tests/test_transport.py]:
```suggestion
```

I don't understand what this test is testing.

File: AGENTS.md [hunk 47]
```diff
@@ -67,3 +67,6 @@ Do NOT edit these directly — modify source scripts instead:
 | `sentry_sdk/profiler/` | Performance profiling |
 | `tests/integrations/{name}/` | Integration test suites |
 | `scripts/populate_tox/config.py` | Test suite configuration |
+
+<!-- This section is maintained by the coding agent via lore (https://github.com/BYK/opencode-lore) -->
+<!-- End lore-managed section -->
```
Comment (sentrivana) [inline AGENTS.md]:
```suggestion
```

**Comment by sentrivana** [comment]
Made a couple changes. I'll take it for a spin in a local project next.

**Comment by sentrivana** [comment]
Added this from local testing: https://github.com/getsentry/sentry-python/pull/5646/changes/4a2a6a3daaf7b4f573d5c492ecfc65bdc50ec873
